domain/data-access/form-repository.ts

```diff
@@ -307,6 +307,147 @@ const saveInstanceVersion = async (
+        OPTIONAL { ?history dct:description ?description }.
+      }
+      ?creator mu:uuid ?creatorId .
+    }
+    ORDER BY DESC(?issued)
+    LIMIT ${limit}
+    OFFSET ${offset}
+  `);
+
+  return result.results.bindings.map((binding) => {
+    return {
+      history: binding.history.value,
+      issued: binding.issued.value,
+      creator: binding.creatorId.value,
+      description: binding.description?.value || null,
+    };
+  });
+};
+
+// unsecure because we don't know if the user has access to the instance
+const unsecureGetInstanceHistoryCount = async (instanceId: string) => {
+  const result = await querySudo(`
+    PREFIX dct: <http://purl.org/dc/terms/>
+    PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
+    PREFIX form: <http://lblod.data.gift/vocabularies/forms/>
+
+    SELECT (COUNT(DISTINCT ?history) AS ?count)
+    WHERE {
+      ?instance mu:uuid ${sparqlEscapeString(instanceId)} .
+      GRAPH <http://mu.semte.ch/graphs/formHistory> {
+        ?history dct:isVersionOf ?instance .
+      }
+    }
+  `);
+
+  const firstResult = result.results.bindings[0];
+  return firstResult?.count?.value || 0;
+};
+
+// unsecure because we don't know if the user has access to the instance
+const unsecureGetHistoryInstance = async (historyUri: string) => {
+  const result = await querySudo(`
+    CONSTRUCT {
+      ?s ?p ?o
+    }
+    WHERE {
+      GRAPH ${sparqlEscapeUri(historyUri)} {
+        ?s ?p ?o
+      }
+    }
+  `);
+
+  return result.results.bindings.map(bindingToTriple).join('\n');
+};
+
+const hasAccessToInstance = async (instanceUri: string) => {
+  const result = await query(`
+    PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
+
+    SELECT * {
+      ${sparqlEscapeUri(instanceUri)} a ?thing .
+    } LIMIT 1
+  `);
+
+  return result.results.bindings.length > 0;
+};
+
+const hasAccessToInstanceId = async (instanceId: string) => {
+  const result = await query(`
+    PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
+
+    SELECT * {
+      ?thing mu:uuid ${sparqlEscapeString(instanceId)}.
+    } LIMIT 1
+  `);
+
+  return result.results.bindings.length > 0;
+};
+
+const hasAccessToHistoryInstance = async (historyInstanceUri: string) => {
+  const instanceResult = await querySudo(`
+    PREFIX dct: <http://purl.org/dc/terms/>
+
+    SELECT ?instance
+    WHERE {
+      GRAPH <http://mu.semte.ch/graphs/formHistory> {
+        ${sparqlEscapeUri(historyInstanceUri)} dct:isVersionOf ?instance .
+      }
+    } LIMIT 1
+  `);
+
+  const instanceUri = instanceResult.results.bindings[0]?.instance?.value;
+
+  if (!instanceUri) {
+    return false;
+  }
+  return hasAccessToInstance(instanceUri);
+};
+
+const getInstanceHistoryWithCount = async (
+  instanceId: string,
+  options: { limit: number; offset: number },
+) => {
+  const [hasAccess, count] = await Promise.all([
+    hasAccessToInstanceId(instanceId),
+    unsecureGetInstanceHistoryCount(instanceId),
+  ]);
+  if (!hasAccess) {
```

Comment:
Might be a bit confusing, wouldn't it be better to actually return some usefull http error code, such as 403?